src/autoscaler/metricsforwarder/server/custom_metrics_handlers.go

```diff
@@ -0,0 +1,206 @@
+package server
+
+import (
+	"autoscaler/db"
+	"autoscaler/metricsforwarder/forwarder"
+	"autoscaler/models"
+	"encoding/json"
+	"errors"
+	"io/ioutil"
+	"net/http"
+	"time"
+
+	"code.cloudfoundry.org/cfhttp/handlers"
+	"code.cloudfoundry.org/lager"
+	cache "github.com/patrickmn/go-cache"
+	"golang.org/x/crypto/bcrypt"
+)
+
+type CustomMetricsHandler struct {
+	metricForwarder    forwarder.MetricForwarder
+	policyDB           db.PolicyDB
+	credentialCache    cache.Cache
+	allowedMetricCache cache.Cache
+	cacheTTL           time.Duration
+	logger             lager.Logger
+}
+
+func NewCustomMetricsHandler(logger lager.Logger, metricForwarder forwarder.MetricForwarder, policyDB db.PolicyDB, credentialCache cache.Cache, allowedMetricCache cache.Cache, cacheTTL time.Duration) *CustomMetricsHandler {
+	return &CustomMetricsHandler{
+		metricForwarder:    metricForwarder,
+		policyDB:           policyDB,
+		credentialCache:    credentialCache,
+		allowedMetricCache: allowedMetricCache,
+		cacheTTL:           cacheTTL,
+		logger:             logger,
+	}
+}
+
+func (mh *CustomMetricsHandler) PublishMetrics(w http.ResponseWriter, r *http.Request, vars map[string]string) {
+	w.Header().Set("Content-Type", "application/json")
+
+	username, password, authOK := r.BasicAuth()
+
+	if authOK == false {
+		mh.logger.Info("error-processing-authorizaion-header")
+		handlers.WriteJSONResponse(w, http.StatusUnauthorized, models.ErrorResponse{
+			Code:    "Authorization-Failure-Error",
+			Message: "Error verifying user credentials. Basic authorization is not used properly"})
+		return
+	}
+
+	var isValid bool
+	var credentials models.CustomMetricCredentials
+
+	appID := vars["appid"]
+	res, found := mh.credentialCache.Get(appID)
+	if found {
+		// Credentials found in cache
+		credentials = res.(models.CustomMetricCredentials)
+		isValid = mh.validateCredentials(username, credentials.Username, password, credentials.Password)
+	}
+
+	// Credentials not found in cache or
+	// stale cache entry with invalid credential found in cache
+	// search in the database and update the cache
+	if !found || !isValid {
+		usernameHash, passwordHash, err := mh.policyDB.GetCustomMetricsCreds(appID)
```

Comment:
if the app keeps using the wrong credential,  it will always query database and update cache. I think we can 
1) increase the interval to active cache update
2) update the cache passively when last update time is long enough